aws/src/main/java/org/apache/iceberg/aws/glue/LockManagers.java

```diff
@@ -0,0 +1,250 @@
+
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import org.apache.iceberg.CatalogProperties;
+import org.apache.iceberg.common.DynConstructors;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.MoreExecutors;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.Tasks;
+
+class LockManagers {
+
+  private static final LockManager LOCK_MANAGER_DEFAULT = new InMemoryLockManager(Maps.newHashMap());
+
+  private LockManagers() {
+  }
+
+  public static LockManager defaultLockManager() {
+    return LOCK_MANAGER_DEFAULT;
+  }
+
+  public static LockManager from(Map<String, String> properties) {
+    if (properties.containsKey(CatalogProperties.LOCK_IMPL)) {
+      return loadLockManager(properties.get(CatalogProperties.LOCK_IMPL), properties);
+    } else {
+      return defaultLockManager();
+    }
+  }
+
+  private static LockManager loadLockManager(String impl, Map<String, String> properties) {
+    DynConstructors.Ctor<LockManager> ctor;
+    try {
+      ctor = DynConstructors.builder(LockManager.class).hiddenImpl(impl).buildChecked();
+    } catch (NoSuchMethodException e) {
+      throw new IllegalArgumentException(String.format(
+          "Cannot initialize LockManager, missing no-arg constructor: %s", impl), e);
+    }
+
+    LockManager lockManager;
+    try {
+      lockManager = ctor.newInstance();
+    } catch (ClassCastException e) {
+      throw new IllegalArgumentException(
+          String.format("Cannot initialize LockManager, %s does not implement LockManager.", impl), e);
+    }
+
+    lockManager.initialize(properties);
+    return lockManager;
+  }
+
+  abstract static class BaseLockManager implements LockManager {
+
+    private static volatile ScheduledExecutorService scheduler;
+
+    private long acquireTimeoutMs;
+    private long acquireIntervalMs;
+    private long heartbeatIntervalMs;
+    private long heartbeatTimeoutMs;
+    private int heartbeatThreads;
+
+    public long heartbeatTimeoutMs() {
+      return heartbeatTimeoutMs;
+    }
+
+    public long heartbeatIntervalMs() {
+      return heartbeatIntervalMs;
+    }
+
+    public long acquireIntervalMs() {
+      return acquireIntervalMs;
+    }
+
+    public long acquireTimeoutMs() {
+      return acquireTimeoutMs;
+    }
+
+    public int heartbeatThreads() {
+      return heartbeatThreads;
+    }
+
+    @SuppressWarnings("StaticGuardedByInstance")
+    public ScheduledExecutorService scheduler() {
+      if (scheduler == null) {
+        synchronized (this) {
```

Comment:
Because this isn't a static method, `this` is the instance of the lock manager, so it doesn't really protect creation of the thread pool. I think you'd need to lock on `BaseLockManager.class` or some other static shared object.

Reply:
Fixed